### PR TITLE
Decorrelate IN_SUBQUERY refactor

### PR DESCRIPTION
**Pull request by max-hoffman**
Decorrelating IN_SUBQUERY into subquery aliases adds new relations to the query plans in a way that makes table and column identifier tracking difficult. So instead of converting `select * from xy where x in (select u from uv)` into a subquery alias join:
```
Filter(
  Tablescan(xy),
  InSubquery(
    x,
    Subquery(
      Project(u, Tablescan(uv))))))
=>
Join(
  (x = u),
  Tablescan(xy),
  SubqueryAlias(
    'scalarSubq0',
    Tablescan(uv)))
```
we disambiguate the subquery and create a table-table join:

```
Filter(
  Tablescan(xy),
  InSubquery(
    x,
    Subquery(
      Project(u, Tablescan(uv))))))
=>
Join(
  (x = u),
  Tablescan(xy),
  Tablescan(uv))
```

We disambiguate table names and column references in the process of unnesting scopes to avoid table name clashes.

This change is better in most places, but worse in instances when we cannot unnest now because the RHS equality expression is not a valid get field reference, for example when the subquery returns the value of a GROUP_BY or WINDOW or a synthesized alias (anything without a source column Id). This is fixable in the future.

This also fixes a few ANTI_JOIN problems that this refactor exposed.

### CODE REVIEW

File: sql/analyzer/unnest_insubqueries.go [hunk 1]
```diff
@@ -122,30 +127,29 @@ func transformJoinApply(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.S
...
-				sch := subq.Query.Schema()
-				var rightF sql.Expression
-				if len(sch) == 1 {
-					subqCol := subq.Query.Schema()[0]
-					rightF = expression.NewGetFieldWithTable(len(scope.Schema()), subqCol.Type, subqCol.DatabaseSource, name, subqCol.Name, subqCol.Nullable)
-				} else {
-					tup := make(expression.Tuple, len(sch))
-					for i, c := range sch {
-						tup[i] = expression.NewGetFieldWithTable(len(scope.Schema())+i, c.Type, c.DatabaseSource, name, c.Name, c.Nullable)
+				if aliases == nil {
```
Comment (JCOR11599) [inline sql/analyzer/unnest_insubqueries.go:130]: I think you can move this block out of the for loop, and remove the `nil` check.

**Review by JCOR11599 — APPROVED** [review]
LGTM

File: enginetest/queries/query_plans.go [hunk 2]
```diff
@@ -129,23 +127,23 @@ From xy;`,
 			"         ├─ columns: [xy.x:1!null, xy.y:2]\n" +
 			"         └─ LookupJoin\n" +
 			"             ├─ Distinct\n" +
-			"             │   └─ Project\n" +
-			"             │       ├─ columns: [scalarSubq0.v:1]\n" +
-			"             │       └─ TableAlias(scalarSubq0)\n" +
+			"             │   └─ Distinct\n" +
```
Comment (zachmu) [inline enginetest/queries/query_plans.go:130]: Double distinct seems sub-optimal, have a plan to fix?

File: enginetest/queries/query_plans.go [hunk 3]
```diff
@@ -154,18 +152,18 @@ From xy;`,
 			"     └─ LookupJoin\n" +
 			"         ├─ Eq\n" +
 			"         │   ├─ xy.y:2\n" +
-			"         │   └─ scalarSubq0.u:0!null\n" +
-			"         ├─ OrderedDistinct\n" +
-			"         │   └─ Project\n" +
-			"         │       ├─ columns: [scalarSubq0.u:0!null]\n" +
-			"         │       └─ TableAlias(scalarSubq0)\n" +
+			"         │   └─ uv.u:0!null\n" +
+			"         ├─ Distinct\n" +
```
Comment (zachmu) [inline enginetest/queries/query_plans.go:156]: Same comment, need to collapse these redundant nodes

File: sql/analyzer/unnest_insubqueries.go [hunk 4]
```diff
@@ -156,70 +160,127 @@ func transformJoinApply(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.S
...
-		ret := tab.WithName(sq.Name())
-		if len(filters) > 0 {
-			filters, err := renameAliasesInExpressions(filters, tab.Name(), sq.Name())
+	for k, v := range duplicates {
+		used[k] = v
+	}
+	return n, err
+}
+
+func getHighestProjection(rec sql.Node) (sql.Expression, bool, error) {
```
Comment (zachmu) [inline sql/analyzer/unnest_insubqueries.go:215]: Need a method doc, summary and return types

Comment (zachmu) [reply sql/analyzer/unnest_insubqueries.go:215]: Also what is `rec`, choose a better name

File: sql/analyzer/unnest_insubqueries.go [hunk 5]
```diff
@@ -156,70 +160,127 @@ func transformJoinApply(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.S
...
+			case expression.Tuple:
+				proj = append(proj, e.Children()...)
+			default:
+				proj = append(proj, e)
+			}
+		case *plan.GroupBy:
+			proj = n.SelectedExprs
+		case *plan.Window:
+			proj = n.SelectExprs
+			// TODO tableIdNode
```
Comment (zachmu) [inline sql/analyzer/unnest_insubqueries.go:256]: Relevant?

File: sql/memo/join_order_builder.go [hunk 6]
```diff
@@ -147,6 +148,16 @@ func NewJoinOrderBuilder(memo *Memo) *joinOrderBuilder {
 	}
 }
 
+var ErrUpsupportedReorderNode = errors.New("unsupported join reorder node")
```
Comment (zachmu) [inline sql/memo/join_order_builder.go:151]: Typo

File: sql/plan/empty_table.go [hunk 7]
```diff
@@ -32,17 +32,31 @@ var _ sql.Node = (*EmptyTable)(nil)
...
 var _ sql.DeletableTable = (*EmptyTable)(nil)
+var _ sql.RenameableNode = (*EmptyTable)(nil)
 
 type EmptyTable struct {
 	schema sql.Schema
 }
 
+func (e *EmptyTable) WithName(s string) sql.Node {
+	ret := *e
+	newSch := make(sql.Schema, len(ret.schema))
```
Comment (zachmu) [inline sql/plan/empty_table.go:43]: Probably easier to use schema.Copy() here instead of copying every column

File: sql/rowexec/join_iters.go [hunk 8]
```diff
@@ -306,12 +308,20 @@ func (i *existsIter) Next(ctx *sql.Context) (sql.Row, error) {
...
 				}
 			} else {
+				i.rightIterNonEmpty = true
 				nextState = esCompare
 			}
 		case esRightIterEOF:
 			if i.typ.IsSemi() {
 				// reset iter, no match
 				nextState = esIncLeft
+			} else if !i.rightIterNonEmpty && !isTrueLit(i.cond) {
```
Comment (zachmu) [inline sql/rowexec/join_iters.go:318]: This latter condition seems pretty fragile, at least add a note about why a condition would be a literal true (as opposed to an expression that always evaluates to true, like `1`)

And if this is effectively a marker for a certain node type, consider pulling that out and making this logic explicit

**Review by zachmu — APPROVED** [review]
Mostly looks good, only minor comments about naming and such

File: sql/analyzer/unnest_insubqueries.go (hunk 1 above)
Comment (max-hoffman) [reply sql/analyzer/unnest_insubqueries.go:130]: i want to lazily collect aliases

File: enginetest/join_planning_tests.go [hunk 9]
```diff
@@ -312,33 +312,33 @@ var JoinPlanningTests = []struct {
 		},
 		tests: []JoinPlanTest{
 			{
-				q:     "select /*+ LOOKUP_JOIN(scalarSubq0,xy) JOIN_ORDER(scalarSubq0,xy) */ * from xy where x = 1 and y in (select a from ab);",
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply enginetest/join_planning_tests.go:315]: We don't make new relations anymore. If we started decorrelating projection subqueries then we maybe would have to worry about this again, unless we switched to using id's for execution indexing.

File: sql/rowexec/join_iters.go (hunk 8 above)
Comment (max-hoffman) [reply sql/rowexec/join_iters.go:318]: i tried a couple variations but we would have to create a new sentinel expression mode to do this. It's not possible to manually make a NOT EXISTS with anything but the literal 1, and if someone changes that things will break loudly.

File: enginetest/queries/query_plans.go (hunk 2 above)
Comment (max-hoffman) [reply enginetest/queries/query_plans.go]: good catch, added a rule now that this is possible

**Pull request merged**